src/hotspot/share/opto/mulnode.cpp

```diff
@@ -505,6 +505,15 @@ const Type *AndINode::mul_ring( const Type *t0, const Type *t1 ) const {
   return TypeInt::INT;          // No constants to be had
 }
 
+const Type* AndINode::Value(PhaseGVN* phase) const {
+  // patterns similar to (v << 2) & 3
+  if (AndIL_shift_and_mask(phase, in(2), in(1), T_INT)) {
+    return TypeInt::ZERO;
+  }
+
+  return MulINode::Value(phase);
```

Comment:
There is no `MulINode::value()` or `MulLNode::value()`, only based class `MulNode` have it.
You missing initial check done by `MulNode::value()`.
I suggest to move this code into `AndINode::mul_ring()` and `AndLNode::mul_ring()`.
Also in `mul_ring()` you can pass `r1->get_con()` as `mask` value to `AndIL_shift_and_mask()`.

Reply:
Thanks for reviewing this.
MulNode::AndIL_shift_and_mask() needs the shift node so it can test for shift->Opcode() == Op_ConvI2L. mult_ring() only gets the Type* as input. Are you suggesting to change mul_ring's signature?

Reply:
You are right, I missed that.
It is unfortunate - we will have many duplicated checks.